### DISCUSSION TITLE
null_resources 사용하는 좋은 예시가 있을까요?

### DISCUSSION THREAD

**Discussion by lcw729** [reliability-engineering-hub/terraform-up-and-running#15] [Q&A] [answered]
좋은 예시가 있다면 공유 부탁드립니다.

**Comment by zxcv9203** [comment, marked as answer]
`null_resource` 리소스는 Terraform에서 유효한 리소스 유형이 아니지만, 지정된 명령을 수행하는 플랫폼 지정 리소스를 작성하는 데 사용됩니다. `null_resource`는 로컬 실행, 원격 실행 및 소스 코드 제어와 같은 다양한 용도로 사용될 수 있습니다.

다음은 `null_resource`를 사용하여 로컬 셸 명령을 실행하는 예시입니다.

```
resource "null_resource" "example" {
  provisioner "local-exec" {
    command = "echo 'Hello, World!'"
  }
}
```

위 코드에서는 `null_resource`를 사용하여 `provisioner` 블록에 정의된 로컬 셸 명령을 실행합니다. 이 리소스를 적용하면 `echo 'Hello, World!'` 명령이 실행됩니다.

다음은 `null_resource`를 사용하여 테라폼 명령 라이프 사이클에 추가하여 `terraform apply`시 최신화된 테라폼 코드를 git에서 받아올 수 있습니다.

```
resource "null_resource" "example" {
  provisioner "local-exec" {
    command = "git clone https://github.com/example/repo.git"
  }
}
```

위 코드에서는 `null_resource`를 사용하여 `provisioner` 블록에 정의된 `git` 명령을 실행합니다. 이 리소스를 적용하면 지정된 리포지토리를 클론하고 코드를 가져올 수 있습니다.

`null_resource`는 다양한 용도로 사용될 수 있지만, 이러한 리소스를 사용하기 전에는 꼭 필요한 경우인지 확인하고 권장되지 않는 패턴을 사용하지 않도록 주의해야 합니다. 또한 `null_resource`를 사용할 때는 프로비저너를 사용하여 리소스의 상태를 변경하지 않도록 주의해야 합니다.

**Comment by lcw729** [comment]
terraform apply를 실행할 때마다 해당 null_resource를 생성하도록 할 수 있습니다.
하지만, 좋은 방법인지는 더 생각해봐야할 것 같습니다.

```
resource "null_resource" "example" {
  # UUID를 사용하여 이 null_resource를 매번 다시 만들도록 한다.
  # terraform apply를 호출 시
  triggers = {
    uuid = uuid()
  }

  provisioner "local-exec" {
    command = "echo \"Hello, World from $(uname -smp)\""
  }
}
```
